File: data-structures/LinkedList/single-linked-list.go
package main

import "fmt"

type Node struct {
	value    interface{}
	next     *Node
	previous *Node // if list is double linked list
}

type LinkedList struct {
	first *Node
}

func NewLinkedList() *LinkedList {
	return &LinkedList{first: nil}
}

func (ll *LinkedList) AddFirst(value interface{}) {
	node := &Node{value: value}
	node.next = ll.first
	ll.first = node
}

func (ll *LinkedList) AddLast(value interface{}) {
	node := &Node{value: value}
	if ll.first != nil {
		currentNode := ll.first
		for currentNode.next != nil {
			currentNode = currentNode.next
		}
		currentNode.next = node
	} else {
		ll.first = node
	}
}

func (ll *LinkedList) RemoveLast() *Node {
	var target *Node
	current := ll.first
	if current != nil && current.next != nil {
		for current.next != nil && current.next.next != nil {
			current = current.next
		}
		target = current.next
		current.next = nil
	} else {
		target = ll.first
		ll.first = nil
	}
	return target
}

func (ll *LinkedList) RemoveFirst() interface{} {
	if ll.first != nil {
		first := ll.first
		ll.first = first.next
		return first.value
	}
	return nil
}

func (ll *LinkedList) Contains(value interface{}) int {
	index := 0
	for current := ll.first; current != nil; current = current.next {
		if current.value == value {
			return index
		}
		index++
	}
	return -1
}

func (ll *LinkedList) RemoveAt(nth int) interface{} {
	if nth == 0 {
		return ll.RemoveFirst()
	}
	current := ll.first
	for index := 0; current != nil; index++ {
		if index == nth-1 {
			if current.next == nil {
				return ll.RemoveLast().value
			}
			removedValue := current.next.value
			current.next = current.next.next
			return removedValue
		}
		current = current.next
	}
	return nil
}

func main() {
	// Example usage
	list := NewLinkedList()
	list.AddFirst(1)
	list.AddLast(2)
	list.AddLast(3)
	fmt.Println(list.Contains(2))
	fmt.Println(list.RemoveAt(1))
	fmt.Println(list.RemoveFirst())
}

File: data-structures/LinkedList/single-linked-list.js
class Node {
  constructor(value = null) {
    this.value = value;
    this.next = null;
    this.previous = null; // if list is double linked list.
  }
}

class LinkedList {
  constructor() {
    this.first = null;
  }

  addFirst(value) {
    const node = new Node(value);
    node.next = this.first;
    this.first = node;
  }

  addLast(value) {
    const node = new Node(value);
    if (this.first) {
      let currentNode = this.first;
      while (currentNode && currentNode.next) {
        currentNode = currentNode.next;
      }
      currentNode.next = node;
    } else {
      this.first = node;
    }
  }

  removeLast() {
    let target = null;
    let current = this.first;
    if (current && current.next) {
      while (current && current.next && current.next.next) {
        current = current.next;
      }
      target = current.next;
      current.next = null;
    } else {
      target = this.first;
    }
    if (target) {
      return target;
    }
  }

  removeFirst() {
    const first = this.first;
    if (first) {
      this.first = first.next;
      return first.value;
    }
  }

  contains(value) {
    let index = 0;
    for (let current = this.first; current; current = current.next) {
      if (current.value === value) {
        return index;
      }
      index++;
    }
    return index;
  }

  removeAt(nth) {
    if (nth === 0) {
      return this.removeFirst;
    }

    for (
      let current = this.first, index = 0;
      current;
      index++, current = current.next
    ) {
      if (index === nth) {
        if (!current.next) {
          return this.removeLast();
        }
        current.previous = current.next;
        return current.value;
      }
    }
  }
}
